ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveRulesRegistry.java

```diff
@@ -27,23 +27,35 @@
 import com.google.common.collect.ListMultimap;
 import com.google.common.collect.SetMultimap;
 import com.google.common.collect.Sets;
+import org.apache.calcite.rel.core.Filter;
 
 public class HiveRulesRegistry {
 
-  private SetMultimap<RelOptRule, RelNode> registryVisited;
-  private ListMultimap<RelNode, Set<String>> registryPushedPredicates;
+  private final SetMultimap<RelOptRule, String> registryVisited;
+  private final ListMultimap<RelNode, Set<String>> registryPushedPredicates;
 
   public HiveRulesRegistry() {
     this.registryVisited = HashMultimap.create();
     this.registryPushedPredicates = ArrayListMultimap.create();
   }
 
   public void registerVisited(RelOptRule rule, RelNode operator) {
-    this.registryVisited.put(rule, operator);
+    this.registryVisited.put(rule, getRelNodeIdentifier(operator));
   }
 
-  public Set<RelNode> getVisited(RelOptRule rule) {
-    return this.registryVisited.get(rule);
+  public boolean hasBeenVisitedBy(RelOptRule rule, RelNode node) {
+    return this.registryVisited.get(rule).contains(getRelNodeIdentifier(node));
+  }
+
+  private String getRelNodeIdentifier(RelNode node) {
+    if (node instanceof Filter) {
+      String inputClass = ((Filter) node).getInput().getClass().getSimpleName();
+      String condition = ((Filter) node).getCondition().toString();
+      String rowType = node.getRowType().toString();
```

Comment:
I think there is an edge case where this check can prevent some optimizations to happen. If you have the same `Filter` in two different places in the tree (e.g., CTEs) then from now on the `HivePreFilteringRule` will trigger only on one of those. I am not sure if this should be a show stopper for merging the PR. I am trying to think of a better way to handle this but don't have many ideas at the moment.